### PR TITLE
Fix #3276: Create Build.bazel file for onboarding

### LINKED ISSUE #3276: Create a BUILD.bazel file for domain/onboarding
This task is part of the build system migration from Gradle to Bazel. 

## Prerequesites

- Make sure you are able to build Oppia using Bazel
- Follow the [Oppia Bazel Setup instructions](https://github.com/oppia/oppia-android/wiki/Oppia-Bazel-Setup-Instructions) to setup Bazel and learn the terminology.
- Read the [Gradle Bazel Migration Best Practies and FAQ](https://github.com/oppia/oppia-android/wiki/Gradle--Bazel-Migration-Best-Practices-and-FAQ) for more context and see if there's an answer there to any question you have.

## Task

1. Create a BUILD.bazel file in

`domain/src/main/java/org/oppia/android/domain/onboarding/`

and

`domain/src/main/java/org/oppia/android/domain/onboarding/testing`

2. Define the kt_android_library library/libraries

Based on the imports in the Kotlin files, add the required third_party dependencies to make the libraries build. Check the Useful links section for a list of third_party dependencies.  You may also need to add the non third_party dependencies.

Create libraries to separate interfaces, annotations and implementations. Often we can split a package into the libraries: api, module and interface.

3. Add the directory to the MIGRATED_PROD_FILES of the module's root BUILD.bazel file

4. Add the libraries to the list of deps dependencies in the targets that require the library

You will have to set the apropriate visbility. See the list of visibilities in the link below.

5. Clean up any unused deps dependencies

6. Verify that the app builds

## Useful links

- [BUILD.bazel files](https://github.com/oppia/oppia-android/search?q=filename%3ABUILD.bazel&type=code)
- [Visibilities](https://github.com/oppia/oppia-android/blob/76dc13dba3ad77aecd2218a5372949d6551f12fe/BUILD.bazel)
- [third_party dependencies](https://github.com/oppia/oppia-android/blob/a70d9b016f0d77dd30c4939041f82ea02ee28d1d/third_party/versions.bzl)

### PR DESCRIPTION
**Pull request by rishidyno**
Fix #3276

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

**Comment by rishidyno** [comment]
@BenHenning I am getting this error can you please help me out with this. Even after adding dependencies to the app BUILD.bazel file, I am not able to build below are the log files.
Screenshot from 2021-10-04 22-22-43


Screenshot from 2021-10-04 22-22-51

**Comment by BenHenning** [comment]
Thanks @rishidyno! Apologies, but I'll need to take a look at this tomorrow.

File: domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel]: This should be removed if there are no dependencies for this library.

File: domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel]: Does this need API visibility? Prefer to keep implementations private.

File: domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel]: We shouldn't abbreviate library names since that could cause confusion. Suggest instead: exploration_meta_data_retriever.

Ditto for exp_mdr_impl below (though for that, it's fine just to call it "impl" since that's conventional).

File: domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel]: Should be oppia_prod_module_visibility instead.

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel]: Suggest instead: ``Test-only utilities for onboarding domain components.``

Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel]: Suggest instead: ``Tests for onboarding domain components.``

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel:8]: This & below:
- Please update the names to be typed out
- Please mark them as testonly (i.e. by adding ``testonly = True``)

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel]: Please fix formatting here (``],`` should be on a new line).

File: app/BUILD.bazel
Comment (BenHenning) [inline app/BUILD.bazel]: These are test-only utilities that shouldn't be pulled into a production library. Could they be added to TEST_DEPS, isntead?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @rishidyno. Had some follow-up comments. PTAL.

> @BenHenning I am getting this error can you please help me out with this. Even after adding dependencies to the app BUILD.bazel file, I am not able to build below are the log files. Screenshot from 2021-10-04 22-22-43
> 
> Screenshot from 2021-10-04 22-22-51

This issue is coming because OnboardingSlideFinalViewModel needs AppStartupStateController, but its dependency can't be resolved. You only added AppStartupStateController to "app" library, but OnboardingSlideFinalViewModel is defined as part of "view_models" (which means view_models' deps need to be updated to point to AppStartupStateController's library).

File: domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel
Comment (rishidyno) [reply domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel]: Removed.

File: domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel
Comment (rishidyno) [reply domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel]: Changed.

File: domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel
Comment (rishidyno) [reply domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel]: Noted. 
Done! Also for fake_exp_mdr.

File: domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel
Comment (rishidyno) [reply domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel]: I think no. Changed it to private.

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (rishidyno) [reply domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel]: Changed.

Comment (rishidyno) [reply domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel]: I think you meant it for domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel. If not please update me with the heading for now its 
`Utilities for onboarding domain components.`

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (rishidyno) [reply domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel:8]: Done.

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (rishidyno) [reply domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel]: Done

**Comment by rishidyno** [comment]
> Thanks @rishidyno. Had some follow-up comments. PTAL.
> 
> > @BenHenning I am getting this error can you please help me out with this. Even after adding dependencies to the app BUILD.bazel file, I am not able to build below are the log files. Screenshot from 2021-10-04 22-22-43
> > Screenshot from 2021-10-04 22-22-51
> 
> This issue is coming because OnboardingSlideFinalViewModel needs AppStartupStateController, but its dependency can't be resolved. You only added AppStartupStateController to "app" library, but OnboardingSlideFinalViewModel is defined as part of "view_models" (which means view_models' deps need to be updated to point to AppStartupStateController's library).

Actually, I tried doing that but there is no build.bazel file for app/onboarding. For this I will have to create bazel file for app/onboarding as well. should I do that?

File: app/BUILD.bazel
Comment (rishidyno) [reply app/BUILD.bazel]: Yes. Added it to TEST_DEPS.

**Comment by rishidyno** [comment]
> > Thanks @rishidyno. Had some follow-up comments. PTAL.
> > > @BenHenning I am getting this error can you please help me out with this. Even after adding dependencies to the app BUILD.bazel file, I am not able to build below are the log files. Screenshot from 2021-10-04 22-22-43
> > > Screenshot from 2021-10-04 22-22-51
> > 
> > 
> > This issue is coming because OnboardingSlideFinalViewModel needs AppStartupStateController, but its dependency can't be resolved. You only added AppStartupStateController to "app" library, but OnboardingSlideFinalViewModel is defined as part of "view_models" (which means view_models' deps need to be updated to point to AppStartupStateController's library).
> 
> Actually, I tried doing that but there is no build.bazel file for app/onboarding. For this I will have to create bazel file for app/onboarding as well. should I do that?

Please, let me know If I have left any changes suggested by you.

File: domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel]: Prefer just "impl" per my earlier comment.

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel:9]: I suggest updating the prod library version of this module's name to be ``retriever_prod_module`` for parity.

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel]: The current code is actually what I meant; my suggestion was incomplete. Thanks!

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @rishidyno. Had some more follow-up comments--PTAL.

> > > Thanks @rishidyno. Had some follow-up comments. PTAL.
> > > > @BenHenning I am getting this error can you please help me out with this. Even after adding dependencies to the app BUILD.bazel file, I am not able to build below are the log files. *screenshot*
> > > > *screenshot*
> > > 
> > > 
> > > This issue is coming because OnboardingSlideFinalViewModel needs AppStartupStateController, but its dependency can't be resolved. You only added AppStartupStateController to "app" library, but OnboardingSlideFinalViewModel is defined as part of "view_models" (which means view_models' deps need to be updated to point to AppStartupStateController's library).
> > 
> > 
> > Actually, I tried doing that but there is no build.bazel file for app/onboarding. For this I will have to create bazel file for app/onboarding as well. should I do that?
> 
> Please, let me know If I have left any changes suggested by you.

You'll need to change https://github.com/oppia/oppia-android/blob/2c6aebd91635ab8dd31488d60aec987c6c37b811/app/BUILD.bazel#L636 sincce that's the library that contains the view model that needs the new dependency.

File: domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel
Comment (rishidyno) [reply domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel]: Done.

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (rishidyno) [reply domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel:9]: Done.

**Comment by rishidyno** [comment]
@BenHenning Made all required changes PTAL.

**Comment by rishidyno** [comment]
@BenHenning I worked on this issue and finally BAZEL build was successful on my local machine but a few checks are failing which I tried but could not understand.

These are the logs. Please have a look.

WARNING: errors encountered while analyzing target '//domain:src/test/java/org/oppia/android/domain/classify/rules/dragAndDropSortInput/DragDropSortInputHasElementXAtPositionYRuleClassifierProviderTest': it will not be built
WARNING: errors encountered while analyzing target '//domain:src/test/java/org/oppia/android/domain/classify/rules/numberwithunits/NumberWithUnitsIsEqualToRuleClassifierProviderTest': it will not be built
WARNING: errors encountered while analyzing target '//domain:src/test/java/org/oppia/android/domain/classify/rules/numericinput/NumericInputIsLessThanRuleClassifierProviderTest': it will not be built
WARNING: errors encountered while analyzing target '//domain:src/test/java/org/oppia/android/domain/classify/rules/ratioinput/RatioInputHasNumberOfTermsEqualToClassifierProviderTest': it will not be built
INFO: Analysis succeeded for only 2 of 28 top-level targets
INFO: Analyzed 28 targets (0 packages loaded, 0 targets configured).
INFO: Found 2 targets...
[0 / 1] [Prepa] BazelWorkspaceStatusAction stable-status.txt
ERROR: command succeeded, but not all targets were analyzed
INFO: Elapsed time: 0.261s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
Error: Process completed with exit code 1.

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel]: This was actually correct before--it should be 'retriever_test_module' since it is a test module.

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel:9]: Sorry to clarify, this means that the _other_ kt_android_library (``retriever_module``) should be updated to ``retriever_prod_module``, not this one.

**Review by BenHenning — COMMENTED** [review]
Thanks @rishidyno. Some CI checks are still failing, but I suspect that this is partly because you just recently got Bazel working locally for you (that's really great to hear, by the way!). Please look into fixing these after we workaround the error you mentioned.

Besides that I had just one comment.

Regarding your error:

> @BenHenning I worked on this issue and finally BAZEL build was successful on my local machine but a few checks are failing which I tried but could not understand.
> 
> These are the logs. Please have a look.
> 
> WARNING: errors encountered while analyzing target '//domain:src/test/java/org/oppia/android/domain/classify/rules/dragAndDropSortInput/DragDropSortInputHasElementXAtPositionYRuleClassifierProviderTest': it will not be built WARNING: errors encountered while analyzing target '//domain:src/test/java/org/oppia/android/domain/classify/rules/numberwithunits/NumberWithUnitsIsEqualToRuleClassifierProviderTest': it will not be built WARNING: errors encountered while analyzing target '//domain:src/test/java/org/oppia/android/domain/classify/rules/numericinput/NumericInputIsLessThanRuleClassifierProviderTest': it will not be built WARNING: errors encountered while analyzing target '//domain:src/test/java/org/oppia/android/domain/classify/rules/ratioinput/RatioInputHasNumberOfTermsEqualToClassifierProviderTest': it will not be built INFO: Analysis succeeded for only 2 of 28 top-level targets INFO: Analyzed 28 targets (0 packages loaded, 0 targets configured). INFO: Found 2 targets... [0 / 1] [Prepa] BazelWorkspaceStatusAction stable-status.txt ERROR: command succeeded, but not all targets were analyzed INFO: Elapsed time: 0.261s, Critical Path: 0.00s INFO: 1 process: 1 internal. FAILED: Build did NOT complete successfully FAILED: Build did NOT complete successfully Error: Process completed with exit code 1.

I think you're missing part of the error log. Is that everything you see from entering the Bazel command? If you can, please create & send a Gist of the full command output plus the Bazel command that you ran so that I can see the full context.

**Comment by rishidyno** [comment]
@BenHenning I am really sorry for being away.
PTAL
Now, 4 tests are failing.
1.https://github.com/oppia/oppia-android/runs/4261785618?check_suite_focus=true
2.https://github.com/oppia/oppia-android/runs/4261785923?check_suite_focus=true
These are the links to the log.
This test is failing and I can't seem to understand it. 
3. Lint test - which I think I have looked for any mistakes I might have made but I could not find any, please take look into this also, 
I would like to know if there exists a script for auto-formatting of bazel files.

**Comment by BenHenning** [comment]
FYI I'm listed as a codeowner for this PR & I'll be unavailable to perform code reviews over the next 2 weeks--thanks for your patience.

**Comment by BenHenning** [comment]
> @BenHenning I am really sorry for being away. PTAL Now, 4 tests are failing. 1.https://github.com/oppia/oppia-android/runs/4261785618?check_suite_focus=true 2.https://github.com/oppia/oppia-android/runs/4261785923?check_suite_focus=true These are the links to the log. This test is failing and I can't seem to understand it. 3. Lint test - which I think I have looked for any mistakes I might have made but I could not find any, please take look into this also, I would like to know if there exists a script for auto-formatting of bazel files.

@rishidyno the Bazel lint check output isn't very helpful, but you're right that there are syntax errors. Buildifier is a tool that can automatically fix most Bazel build file errors: https://github.com/bazelbuild/buildtools/blob/master/buildifier/README.md.

Regarding the actual failures, the top error I see is:
```
2021-11-19T08:25:54.2670669Z WARNING: errors encountered while analyzing target '//domain:src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest': it will not be built
```

This usually means a syntax or logical build graph issue in the BUILD file the test is defined. Since this is an unmigrated test, that means domain/BUILD.bazel. However, since you didn't change that file then it means one of the build files _it_ depends on has the issue. I suggest going through the build files one-by-one and comment out your changes until you find the culprit line. That should help you then figure out what went wrong.

File: domain/BUILD.bazel
Comment (FareesHussain) [inline domain/BUILD.bazel]:
```suggestion
    "//domain/src/main/java/org/oppia/android/domain/onboarding:state_controller",
```

Comment (rishidyno) [reply domain/BUILD.bazel]: Done

**Comment by rishidyno** [comment]
@BenHenning all tests have passed PTAL .

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel:9]: This still appears to be unresolved.

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel]: This has been addressed.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @rishidyno! I think you might have missed one of my past comments--PTAL (in general, please make sure to reply to all reviewer comments before sending the PR back for review just to make sure everything's been acknowledge or addressed). You can quickly find these by going to the 'Conversations' drop down on the 'Files changed' tab.

Beyond that, the PR looks really good. I think just the one change I'm requesting is needed before this PR can be merged.

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (rishidyno) [reply domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel:9]: @BenHenning I would like to clarify that you want me to change the kt_android_library name from `retriever_module` which is present in `domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel` file to `retriever_prod_module`?

Comment (rishidyno) [reply domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel:9]: @BenHenning changes were addressed PTAL.

**Comment by rishidyno** [comment]
@BenHenning All tests have passed PTAL.

File: domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/onboarding/testing/BUILD.bazel:9]: Yep, that's what I intended @rishidyno. Thanks!

**Review by BenHenning — APPROVED** [review]
Thanks @rishidyno! This LGTM.

**Pull request merged**
